modin/pandas/test/test_groupby.py

```diff
@@ -2099,3 +2100,107 @@ def test_mean_with_datetime(by_func):
 
     modin_df, pandas_df = create_test_dfs(data)
     eval_general(modin_df, pandas_df, lambda df: df.groupby(by=by_func(df)).mean())
+
+
+@pytest.mark.parametrize(
+    "modin_df_recipe",
+    ["non_lazy_frame", "frame_with_deferred_index", "lazy_frame"],
+)
+def test_groupby_on_empty_data(modin_df_recipe):
+    class ModinDfConstructor:
+        def __init__(self, recipie, df_kwargs):
+            self._recipie = recipie
+            self._mock_obj = None
+            self._df_kwargs = df_kwargs
+
+        def non_lazy_frame(self):
+            return pd.DataFrame(**self._df_kwargs)
+
+        def frame_with_deferred_index(self):
+            df = pd.DataFrame(**self._df_kwargs)
+            try:
+                # The frame would stop being lazy once index computation is triggered
+                df._query_compiler._modin_frame._index_cache = None
+            except AttributeError:
+                pytest.skip(
+                    reason="Selected execution doesn't support deferred indices."
+                )
+
+            return df
+
+        def lazy_frame(self):
+            donor_obj = pd.DataFrame()._query_compiler
+
+            self._mock_obj = mock.patch(
+                f"{donor_obj.__module__}.{donor_obj.__class__.__name__}.lazy_execution",
+                new_callable=mock.PropertyMock,
+            )
+            patch_obj = self._mock_obj.__enter__()
+            patch_obj.return_value = True
+
+            df = pd.DataFrame(**self._df_kwargs)
```

Comment:
Isn't it enough just to patch the instance `lazy_execution` attribute? Then you wouldn't have to call `__enter__` and `__exit__` methods.

```suggestion
            df = pd.DataFrame(**self._df_kwargs)
            df._query_compiler = Mock()  # Mock object that return `True`
```

Reply:
Having `lazy_execution=True` set for all of the instances makes us secure that all of the derivatives of `df` would also be lazy, so we wouldn't accidentally test a non-lazy case because some preliminary groupby function replaced the patched `df` with some of its derivatives.

Additionally, [`mock.patch.object`](https://docs.python.org/3/library/unittest.mock.html#patch-object) doesn't work with properties (probably because of their read-only nature):
```python
import modin.pandas as pd
from unittest import mock

md_res = pd.DataFrame(columns=["a", "b", "c"])

with mock.patch.object(
    target=md_res._query_compiler,
    attribute="lazy_execution",
    new_callable=mock.PropertyMock,
) as patch: # AttributeError: can't delete attribute
    patch.return_value = True
    print(md_res._query_compiler.lazy_execution)

with mock.patch.object(
    target=md_res._query_compiler,
    attribute="lazy_execution",
    new=True,
) as patch: # AttributeError: can't delete attribute
    print(md_res._query_compiler.lazy_execution)
```
So I believe that it's better to leave it as it is.

Reply:
ok